### PR TITLE
alts: Major rewrite of the ALTS transport socket to fix concurrent connection establishment bugs (version 2).

### PR DESCRIPTION
**Pull request by matthewstevenson88**
Commit Message:
Additional Description: The existing ALTS transport socket extension cannot handle establishing multiple concurrent connections. This is due to a fundamental problem in how the transport socket is architected on top of the ALTS TSI implementation in the gRPC C-core; specifically, this issue occurs because the ALTS transport socket uses gRPC-internal APIs to try to replicate the gRPC event queue, but fails to do so correctly. To avoid this problem and problems like it in the future, this PR rewrites the ALTS transport socket so that it only depends on public gRPC APIs. We have confirmed that this fixes the original bug and the transport socket can now successfully establish multiple concurrent connections.
Risk Level:
Testing:
Docs Changes:
Release Notes: Rewrite of the ALTS transport socket to fix concurrent connection establishment bugs. The ALTS transport socket can now successfully establish multiple concurrent connections.
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by botengyao** [comment]
Hi @alyssawilk, feel free to assign the ALTS review to me. @matthewstevenson88, we can sync up offline if you need some help. From my last check, we need to fix some existing tests.

File: source/extensions/transport_sockets/alts/tsi_socket.cc
Comment (botengyao) [reply source/extensions/transport_sockets/alts/tsi_socket.cc:318]: besides, is there an integration test cover this?

File: source/extensions/transport_sockets/alts/alts_channel_pool.cc
Comment (botengyao) [inline source/extensions/transport_sockets/alts/alts_channel_pool.cc:24]: Maybe move this to be a const static number right now.

File: source/extensions/transport_sockets/alts/alts_proxy.cc
Comment (botengyao) [inline source/extensions/transport_sockets/alts/alts_proxy.cc:65]: maybe s/:/,

**Review by botengyao — COMMENTED** [review]
Thanks for working on this, and this looks great!
I haven't took a detailed look for the test, but overall the logic LGTM.

**Comment by botengyao** [comment]
Besides, feel free to remove the `[WIP]` marker. and @envoyproxy/envoy-maintainers for the CI approval. Thanks!

**Comment by jmarantz** [comment]
@matthewstevenson88 if this is ready for final review please remove "[WIP]" from the PR title. Thanlk you!

/wait

File: source/extensions/transport_sockets/alts/alts_proxy.cc
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/alts_proxy.cc]: Done.

File: source/extensions/transport_sockets/alts/alts_tsi_handshaker.h
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/alts_tsi_handshaker.h:17]: It is, yes.

File: source/extensions/transport_sockets/alts/config.cc
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/config.cc:106]: Good observation! We have the nullity check where the channel is ultimately used, in the `AltsProxy::create` method. We could also add nullity checks at each stage of the plumbing from here to there, but I think having just the one check should be fine. WDYT?

File: source/extensions/transport_sockets/alts/tsi_socket.cc
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/tsi_socket.cc]: Good call, removed this since it is no longer relevant.

File: source/extensions/transport_sockets/alts/tsi_socket.cc
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/tsi_socket.cc]: Fixed.

AFAIK, there is no integration test in OSS Envoy that caught this. If you can point me to an example to base of, I'm happy to add one (or maybe we should defer this to a follow-up PR since this is an existing issue).

File: source/extensions/transport_sockets/alts/alts_channel_pool.cc
Comment (KBaichoo) [inline source/extensions/transport_sockets/alts/alts_channel_pool.cc:22]: likewise

File: source/extensions/transport_sockets/alts/alts_channel_pool.cc
Comment (KBaichoo) [inline source/extensions/transport_sockets/alts/alts_channel_pool.cc:42]: Creating a BitGen to dtor everytime seems expensive.  Do we need uniform randomness among channels, is round robin good enough? Can we use a SharedBit gen member?

**Review by KBaichoo — COMMENTED** [review]
Thank you for working on this. Here's a quick pass.

Can you add yourself as a codeowner https://github.com/envoyproxy/envoy/blob/620091ac80a3611f3340d21ff558666a9314d465/CODEOWNERS#L35 since this is a huge rewrite and you probably have the best expertise?

Can you add release notes? and add more details in the PR description.

Thanks!

/wait

File: source/extensions/transport_sockets/alts/tsi_socket.cc
Comment (botengyao) [reply source/extensions/transport_sockets/alts/tsi_socket.cc]: I am thinking like we should trigger the assertion if this line is triggered. We can leverage https://github.com/envoyproxy/envoy/blob/47afc8c884ca95a40d01513cf82f0f40b67b8ed8/test/extensions/transport_sockets/alts/alts_integration_test.cc#L195
to run into this condition during `doWrite` call.

File: source/extensions/transport_sockets/alts/alts_tsi_handshaker.h
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/alts_tsi_handshaker.h]: Switched to `std::vector<uint8_t>`.

File: source/extensions/transport_sockets/alts/alts_channel_pool.cc
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/alts_channel_pool.cc]: Good call, switched to round robin.

File: source/extensions/transport_sockets/alts/alts_tsi_handshaker.h
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/alts_tsi_handshaker.h:28]: Before this PR, the ALTS transport socket relied on an implementation of the TSI interface from the gRPC C-core for the lower-level ALTS details. The tsi_handshaker was an Envoy wrapper over this interface from gRPC. All of the code in config.cc and tsi_socket.{h,cc} is built on top of tsi_handshaker, so I didn't want to entirely discard it in this PR (though I may do so in a follow-up with cleanup). Instead, I created alts_tsi_handshaker, which enables us to drop in the new functionality into the existing tsi_handshaker wrapper (with a few edits). Does that help clarify?

**Comment by matthewstevenson88** [comment]
> Thank you for working on this. Here's a quick pass.
> 
> Can you add yourself as a codeowner
> 
> https://github.com/envoyproxy/envoy/blob/620091ac80a3611f3340d21ff558666a9314d465/CODEOWNERS#L35
> 
> since this is a huge rewrite and you probably have the best expertise?
> Can you add release notes? and add more details in the PR description.
> 
> Thanks!

Thanks for the review, and thanks to @botengyao as well! I added myself as a code owner, added more details in the PR description, and added release notes in the PR description.

File: source/extensions/transport_sockets/alts/alts_channel_pool.h
Comment (KBaichoo) [inline source/extensions/transport_sockets/alts/alts_channel_pool.h:31]: can use `ABSL_GUARDED_BY` macros to annotate what this protects?

File: CODEOWNERS
Comment (KBaichoo) [inline CODEOWNERS:35]: I think you can remove @yangminzhu here.

File: source/extensions/transport_sockets/alts/alts_tsi_handshaker.h
Comment (KBaichoo) [reply source/extensions/transport_sockets/alts/alts_tsi_handshaker.h:28]: Thank you for the context here!

File: source/extensions/transport_sockets/alts/config.cc
Comment (KBaichoo) [inline source/extensions/transport_sockets/alts/config.cc:104]: no need to initialize nullptr as will do this in the default ctor.

File: source/extensions/transport_sockets/alts/alts_tsi_handshaker.h
Comment (KBaichoo) [reply source/extensions/transport_sockets/alts/alts_tsi_handshaker.h:45]: That makes sense then to not call it for "ForTest". But  can you point to where we call it explicitly from a test e.g. such that it helps with coverage (I might have missed it). Otherwise if we just use it via the public interface in test, making it private sgtm.

File: source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc
Comment (KBaichoo) [inline source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc:52]: do you mean to drop the unsigned here?

File: source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc
Comment (KBaichoo) [inline source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc:61]: What do you think about splitting this out into a server / client subclasses which can share some common core but may differ in the implementation here (vs us doing work for both and mixing the logic together with if/else guards).

File: source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc
Comment (KBaichoo) [inline source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc:106]: std::clamp will do what you're doing here.

File: test/extensions/transport_sockets/alts/alts_integration_test.cc
Comment (KBaichoo) [inline test/extensions/transport_sockets/alts/alts_integration_test.cc:108]: Do we an an integration test where we concurrently try to make connections (say 2) and we have 2 workers? I think that would help give confidence in fixing the underlying issue (and not introducing another race)

Comment (matthewstevenson88) [reply test/extensions/transport_sockets/alts/alts_integration_test.cc:108]: I have an integration test where I run many concurrent connections through an Envoy running on GCP using the real ALTS handshaker service - this is what I used to confirm the fix.

At present, it is not clear to me how to set up a full integration test in the Envoy repo that does many concurrent connections with a fake ALTS handshaker service. However, I set up unit tests at the TSI handshaker layer that do concurrent handshakes, and these also would have caught the original issue.

File: source/extensions/transport_sockets/alts/tsi_socket.cc
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/tsi_socket.cc]: Sorry I missed this. There are tests in tsi_socket_test.cc that check this code block now.

File: source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc]: Fixed, replaced with `absl::Span<const uint8_t>`.

File: source/extensions/transport_sockets/alts/alts_proxy.h
Comment (KBaichoo) [inline source/extensions/transport_sockets/alts/alts_proxy.h:46]: What is being blocked here? e.g. are we blocking the Envoy silo here with any / all of the calls above? If so we can easily bring the system to a halt by the calls above.

File: source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc
Comment (KBaichoo) [reply source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc:62]: sg, thanks for your consideration.

File: test/extensions/transport_sockets/alts/alts_integration_test.cc
Comment (KBaichoo) [inline test/extensions/transport_sockets/alts/alts_integration_test.cc:148]: prefer enum class

**Review by KBaichoo — COMMENTED** [review]
Thanks for working on this, generally looks good apart from the "blocking" question

/wait

File: test/extensions/transport_sockets/alts/alts_integration_test.cc
Comment (matthewstevenson88) [reply test/extensions/transport_sockets/alts/alts_integration_test.cc]: Done.

File: source/extensions/transport_sockets/alts/alts_proxy.h
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/alts_proxy.h]: No sorry, I just mean that we're blocking the current thread. I removed this comment because I don't think it adds anything.

**Comment by matthewstevenson88** [comment]
> Thanks for working on this, generally looks good apart from the "blocking" question
> 
> /wait

Thanks for your review @KBaichoo - your comments/suggestions are much appreciated. :)

@botengyao Could you also take a final pass when you have a chance?

Thanks both!

File: source/extensions/transport_sockets/alts/alts_proxy.cc
Comment (botengyao) [inline source/extensions/transport_sockets/alts/alts_proxy.cc:39]: Do we need to set_deadline for the gRPC client through the ClientContext, since it is possible to end up waiting forever?

**Review by botengyao — COMMENTED** [review]
Thanks for working on this epic change! LGTM module the package and a question.

/wait

File: source/extensions/transport_sockets/alts/tsi_socket.cc
Comment (botengyao) [reply source/extensions/transport_sockets/alts/tsi_socket.cc]: Thanks!

**Comment by botengyao** [comment]
cc @phlax for CI. Thanks!

File: bazel/repository_locations.bzl
Comment (matthewstevenson88) [reply bazel/repository_locations.bzl]: Done.

File: test/extensions/transport_sockets/alts/alts_proxy_test.cc
Comment (matthewstevenson88) [reply test/extensions/transport_sockets/alts/alts_proxy_test.cc:154]: Done.

File: source/extensions/transport_sockets/alts/alts_proxy.cc
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/alts_proxy.cc:42]: Good idea, done.

**Review by botengyao — COMMENTED** [review]
LGTM, thanks!

@envoyproxy/dependency-shepherds or @RyanTheOptimist for dependency approval and CI.

File: source/extensions/transport_sockets/alts/alts_proxy.cc
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/alts_proxy.cc:29]: Done.

**Comment by adisuissa** [comment]
@KBaichoo any more comments?

**Review by KBaichoo — COMMENTED** [review]
Mostly looks good apart from conversation in https://github.com/envoyproxy/envoy/pull/29803#discussion_r1392930294

File: source/extensions/transport_sockets/alts/alts_proxy.h
Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/alts_proxy.h]: @botengyao to correct me if I explain the TSI socket poorly. :)

When the ALTS transport socket is handling an ALTS handshake, this is occurring in a worker thread, and we are block this worker thread when we call this function and wait for this function to execute. Under the hood of this function, we are making a gRPC call to the ATLS handshaker service.

@KBaichoo Does that help clarify? Can I expand on anything?

Apologies I'm not sure what you mean by "the Envoy silo" above. If this does not answer your question, can you expand on this?

Comment (KBaichoo) [reply source/extensions/transport_sockets/alts/alts_proxy.h]: Thanks for clarifying, as such as can easily bring the entire proxy to a halt if the grpc service is unavailable / experiencing high latency. I think we should strongly document this, and possibly follow up. I suspect mostly Google might use this extension @yanavlasov @adisuissa @botengyao  on if/when this bottleneck would need to be addressed.

Comment (botengyao) [reply source/extensions/transport_sockets/alts/alts_proxy.h]: Right, we should document this. I think the best solution is to make this RPC call as an async stream with response callback, similar as the async gRPC client here. https://github.com/envoyproxy/envoy/blob/1140c93e207bc30995dfb5948bff9e4b48302f6f/source/common/grpc/async_client_impl.h#L12. Did we perform the load test with the current implementation internally with with currently logic?

Comment (matthewstevenson88) [reply source/extensions/transport_sockets/alts/alts_proxy.h]: Thansk @KBaichoo for explaining, really glad you caught this. I've added documentation to alts_proxy.h and added a TODO in alts_proxy.cc to further investigate the async client.

@botengyao I have load testing set up with the internal implementation and I did not notice this issue there. For that reason, I'm tempted to suggest that we merge this PR (so that there is a working version of the ALTS transport socket) and look into a follow-up PR to use the async client. WDYT?

Comment (botengyao) [reply source/extensions/transport_sockets/alts/alts_proxy.h]: sgtm, defer to @KBaichoo for approval. Thanks!

Comment (botengyao) [reply source/extensions/transport_sockets/alts/alts_proxy.h]: Having saying that, and taking a deeper look, I think this still should be the **top priority TODO** to make the gRPC call be async to not block the worker thread based on Envoy's silo. I tend to let this PR merged since there is no actual usage and this makes the extension workable at least, but also glad to hear @KBaichoo and others' opinion.

Comment (KBaichoo) [reply source/extensions/transport_sockets/alts/alts_proxy.h]: +1 to letting this through since I expect there might not be significant usage of this (given the preexisting bug); this is a step in the right direction and this PR is already large enough.

**Review by KBaichoo — APPROVED** [review]

**Comment by adisuissa** [comment]
This is an extension-only modification, so I think !Google review is not required.
Assigning Alyssa for a senior-maintainer stamp.
/assign @alyssawilk

**Pull request merged**
